Civi/Crypto/CryptoToken.php

```diff
@@ -0,0 +1,139 @@
+ */
+
+namespace Civi\Crypto;
+
+use Civi\Crypto\Exception\CryptoException;
+
+/**
+ * The "Crypto Token" service supports a token format suitable for
+ * storing specific values in the database with encryption. Characteristics:
+ *
+ * - Primarily designed to defend confidentiality in case of data-leaks
+ *   (SQL injections, lost backups, etc).
+ * - NOT appropriate for securing data-transmission. Data-transmission
+ *   requires more protections (eg mandatory TTLs + signatures). If you need
+ *   that, consider adding a JWT/JWS/JWE implementation.
+ * - Data-format allows phase-in/phase-out. If you have a datum that was written
+ *   with an old key or with no key, it will still be readable.
+ *
+ * USAGE: The "encrypt()" and "decrypt()" methods are the primary interfaces.
+ *
+ *   $encrypted = Civi::service('crypto.token')->encrypt('my-mail-password, 'KEY_ID_OR_TAG');
+ *   $decrypted = Civi::service('crypto.token')->decrypt($encrypted, '*');
+ *
+ * FORMAT: An encoded token may be in either of these formats:
+ *
+ *   - Plain text: Any string which does not begin with chr(2)
+ *   - Encrypted text: A string in the format:
+ *        TOKEN := DLM + VERSION + DLM + KEY_ID + DLM + CIPHERTEXT
+ *        DLM := ASCII CHAR #2
+ *        VERSION := String, 4-digit, alphanumeric (as in "CTK0")
+ *        KEY_ID := String, alphanumeric and symbols "_-.,:;=+/\"
+ *
+ * @package Civi\Crypto
+ */
+class CryptoToken {
+
+  const VERSION_1 = 'CTK0';
+
+  protected $delim;
+
+  /**
+   * CryptoToken constructor.
+   */
+  public function __construct() {
+    $this->delim = chr(2);
+  }
+
+  /**
+   * Determine if a string looks like plain-text.
+   *
+   * @param string $plainText
+   * @return bool
+   */
+  public function isPlainText($plainText) {
+    return is_string($plainText) && ($plainText === '' || $plainText{0} !== $this->delim);
+  }
+
+  /**
+   * Create an encrypted token (given the plaintext).
+   *
+   * @param string $plainText
+   *   The secret value to encode (e.g. plain-text password).
+   * @param string|string[] $keyIdOrTag
+   *   List of key IDs or key tags to check. First available match wins.
+   * @return string
+   *   A token
+   * @throws \Civi\Crypto\Exception\CryptoException
+   */
+  public function encrypt($plainText, $keyIdOrTag) {
+    /** @var CryptoRegistry $registry */
+    $registry = \Civi::service('crypto.registry');
+
+    $key = $registry->findKey($keyIdOrTag);
+    if ($key['suite'] === 'plain') {
+      if (!$this->isPlainText($plainText)) {
+        throw new CryptoException("Cannot use plaintext encoding for data with reserved delimiter.");
+      }
+      return $plainText;
+    }
+
+    /** @var \Civi\Crypto\CipherSuiteInterface $cipherSuite */
+    $cipherSuite = $registry->findSuite($key['suite']);
+    $cipherText = $cipherSuite->encrypt($plainText, $key);
+    return $this->delim . self::VERSION_1 . $this->delim . $key['id'] . $this->delim . base64_encode($cipherText);
```

Comment:
I'm on the fence about whether to keep the `base64_encode()` for CTK0. It sorta feels like a waste of CPU. OTOH, I think the prospects of reading+writing the token via SQL are better if limit the range of unprintable characters (i.e. $this->delim is an unprintable char - but everything else here is printable).

Reply:
I also wonder if we even care about wanting it SQL readable, I suppose it makes sense but just as a thought but I think what you have put up is a sensible route

Reply:
Haven't tested - but I was expecting to work a bit like the `DAO::VALUE_SEPARATOR`, ie if you use a SQL console and look at a table with an crypted value:

* The delimiter would either appear as funny block, or it would be skipped/hidden.
* The data is otherwise displayable. So you'll see any encrypted values begin with the version header (ie "CTK0").

Reply:
Updated the notation. It now looks like `chr(2)` followed by `CTK?k=<key-id>&t=<cipher-text>`. The two parameters are encoded in [URL-safe Base64](https://tools.ietf.org/html/rfc4648#section-5). Here's an example of how an encrypted token appears in an API call: 

```
[bknix-min:~/bknix/build/dmaster/web/sites/all/modules/civicrm] cv api setting.get return=mailing_backend
{
    "is_error": 0,
    "version": 3,
    "count": 1,
    "id": 1,
    "values": {
        "1": {
            "mailing_backend": {
                "outBound_option": "0",
                "smtpServer": "example.com",
                "smtpPort": "25",
                "smtpAuth": "",
                "smtpUsername": "foo",
                "smtpPassword": "\u0002CTK?k=qrMa7S-_VIa5bKpCxh4A6ppll0A&t=yFkyLXio0QzxzNipLxErPRsyjPx-UxJRlos9Pz7BI6c"
            }
        }
    }
}
```